ledger/apply/challenge.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019-2024 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package apply
+
+import (
+	"math/bits"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/committee"
+)
+
+// ChallengePeriod indicates which part of the challenge period is under discussion.
+type ChallengePeriod int
+
+const (
+	// ChRisky indicates that a challenge is in effect, and the initial grace period is running out.
+	ChRisky ChallengePeriod = iota
+	// ChActive indicates that a challenege is in effect, and the grace period
+	// has run out, so accounts can be suspended
+	ChActive
+)
+
+type challenge struct {
+	// round is when the challenge occurred. 0 means this is not a challenge.
+	round basics.Round
+	// accounts that match the first `bits` of `seed` must propose or heartbeat to stay online
+	seed committee.Seed
+	bits int
+}
+
+type headerSource interface {
+	BlockHdr(round basics.Round) (bookkeeping.BlockHeader, error)
+}
+
+// FindChallenge returns the Challenge that was last issued if it's in the period requested.
+func FindChallenge(rules config.ProposerPayoutRules, current basics.Round, headers headerSource, period ChallengePeriod) challenge {
+	// are challenges active?
+	interval := basics.Round(rules.ChallengeInterval)
+	if rules.ChallengeInterval == 0 || current < interval {
+		return challenge{}
+	}
+	lastChallenge := current - (current % interval)
+	grace := basics.Round(rules.ChallengeGracePeriod)
+	// FindChallenge is structured this way, instead of returning the challenge
+	// and letting the caller determine the period it cares about, to avoid
+	// using BlockHdr unnecessarily.
+	switch period {
+	case ChRisky:
+		if current <= lastChallenge+grace/2 || current > lastChallenge+grace {
+			return challenge{}
+		}
+	case ChActive:
+		if current <= lastChallenge+grace || current > lastChallenge+2*grace {
+			return challenge{}
+		}
+	}
+	challengeHdr, err := headers.BlockHdr(lastChallenge)
+	if err != nil {
+		panic(err)
```

Comment:
It turns out you wouldn't even get an error then, it would pull the block from db.  Which is good, because if we want to verify blocks with suspensions or free heartbeats in them, we need the value.

But I take your point, and I'll return the empty challenge. Which could cause either of the two things above to fail, but that's reasonable, since something is very wrong.